### ISSUE TITLE
Directionality of regional-indicator flags

### ISSUE THREAD

**Issue by arenevier** [harfbuzz/harfbuzz#2265]
If a buffer contains ltr characters needing a ligature, but the script is set to hebrew (which is a rtl script), the characters ligatured are reversed.

To be more clear, if the buffer contains

    |hebrew_characters|regional_indicator_i|regional_indicator_l|

calling hb_shape should result in:

    |hebrew_codepoints|Israel_flag|

Instead, we get:

    |hebrew_codepoints|Liechtenstein_flag|

This happens because of hb_ensure_native_direction is called before the substitution. So during the substitution, L and I are considered as a pair, instead of I and L.

I am not sure if that specific case should be addressed in harfbuzz, or if the caller should address it.

You can see a discussion with more context in chromium bug tracker:
https://bugs.chromium.org/p/chromium/issues/detail?id=1054512

**Comment by khaledhosny** [comment]
This is working as designed. If the script’s native direction is LTR we want, say, `[f|i]` to always form an `[f_i]` ligature regardless of the set direction, and we don’t want `[f|i]` to ever form such a ligature.

IMHO, Chrome setting emoji script to Hebrew is probably a bad idea, and should either set the script to Common or even Latin.

**Comment by khaledhosny** [comment]
Or may be we need to mark regional indicator symbols as part of the same cluster. This quicj patch seems to achieve that:
```diff
iff --git a/src/hb-ot-shape.cc b/src/hb-ot-shape.cc
index 00ecdfab..acd35bb1 100644
--- a/src/hb-ot-shape.cc
+++ b/src/hb-ot-shape.cc
@@ -486,6 +486,17 @@ hb_set_unicode_props (hb_buffer_t *buffer)
        _hb_glyph_info_set_continuation (&info[i]);
       }
     }
+    /* Or a regional indicator symbols sequence. */
+    else if (unlikely (hb_in_range<hb_codepoint_t> (info[i].codepoint, 0x1F1E6u, 0x1F1FFFu)))
+    {
+      if (i + 1 < count &&
+         hb_in_range<hb_codepoint_t> (info[i + 1].codepoint, 0x1F1E6u, 0x1F1FFFu))
+      {
+       i++;
+       _hb_glyph_info_set_unicode_props (&info[i], buffer);
+       _hb_glyph_info_set_continuation (&info[i]);
+      }
+    }
 #endif
     /* Or part of the Other_Grapheme_Extend that is not marks.
      * As of Unicode 11 that is just:
```

**Comment by arenevier** [comment]
Indeed, it fixes the issue.

**Comment by behdad** [comment]
I'm sure we have discussed this with @drott and @roozbehp before.  I don't remember why we didn't do the obvious whch is what Khaled proposed.

I know we have even tried to support subregion tags: https://github.com/harfbuzz/harfbuzz/issues/1556

It's likely that we just missed Regional_Indicators.  Might be just a mistake.  However I have a vague memory that Regional_indicators are bidi=L and as such won't need special handling.  I don't have setup to check that right now.

**Comment by behdad** [comment]
Our implementation is getting pretty good, I like to rip it out and expose it as API.  I have to implement something like libraqm soon.  Will look at raqm and try to come up w script and grapheme iterators we can expose in HarfBuzz.  Am still resisting adding a `hb_string_t` type though.  Not sure what's better.

**Comment by jfkthame** [comment]
Maybe we should seek clarification from Unicode about this before possibly changing behavior. I'm not confident what should be considered "correct".

It's not clear to me whether RI pairs should get special treatment here. If a font *doesn't* have ligatures for them (like Segoe UI Emoji, for instance), then should the two characters be displayed in LTR order even when a directional override is applied? That seems like it would be a violation of the Unicode bidi algorithm.

Consider https://codepen.io/jfkthame/pen/xxGjWzB?editors=1100, where the encoded text in all the lines consists of "Test: [A][U]", where [A] and [U] represent regional-indicator codepoints.

In LTR context with the Apple emoji font (includes flag ligatures), this renders as "Test [Australia-flag]". Using the Segoe font (no flag ligatures), it renders as "Test AU".

Setting paragraph direction RTL has no effect, as all the characters still have LTR directionality, so we still get "Test [Australia-flag]" or "Test AU" depending on font support.

However, what should happen if we apply a directional override, forcing RTL directionality on all the characters? Currently, browsers agree on the Apple emoji rendering "[Ukraine-flag] tseT", which seems somewhat counter-intuitive, but I would argue may be correct given the Unicode representation. This is clearer when we consider the Segoe UI rendering "UA tseT", where it's obvious that the RTL override has been properly applied to each of the individual characters `T`, `e`, `s`, `t`, `<space>`, `[A]`, `[U]`, and the text reads entirely (and correctly) "backwards". But if the Segoe UI rendering shows (visually) "UA", it's no longer a surprise that the Apple emoji rendering shows its "[Ukraine-flag]" ligature.

The proposed fix here would change behavior such that the bidi-overridden examples render as (Apple emoji) "[Australia-flag] tseT" and (Segoe UI) "AU tseT". That intuitively seems like a "good" result for the Apple Emoji version, where the Australian flag "character" has been preserved, but it seems wrong for the Segoe UI version, where the two-character sequence "[A][U]" has *not* been bidi-reordered as would be expected due to the direction override.

The current behavior means that a direction override changes the apparent meaning of the text, which initially seems like a bad thing. But then, that's generally true for any overriding of the inherent direction of a sequence of characters: given text that contains [RLO][l][i][v][e][PDF], the rendered text will appear to say "evil".

I'm conflicted.... not sure what is the right answer!

**Comment by kojiishi** [comment]
The discussion at [crbug.com/1054512#c18](https://crbug.com/1054512#c18) discovered that Gecko runs script segmenter for each bidi runs, while Blink runs for the entire paragraph. For this test case, this is reason why the regional indicators are tagged as Hebrew only in Blink.

I'm fine to match this behavior to Gecko, if people here can reach consensus that it is the best way to go. It solves the reported problem, though doesn't really answer the directional override issue Jonathan mentioned.

**Comment by jfkthame** [comment]
Are you sure the diagnosis here is correct -- that it's being tagged with script=hebrew that is causing a problem?

Testing with hb-shape, I don't see the script tag causing any change in ligated result:

    $ ./BUILD/util/hb-shape --font-file NotoColorEmoji.ttf --unicodes 1f1f9,1f1e6
    [gid1417=0+2550]

    $ ./BUILD/util/hb-shape --font-file NotoColorEmoji.ttf --unicodes 1f1f9,1f1e6 --script hebr
    [gid1417=0+2550]

Whereas directionality is significant (which I'm suggesting may be correct, though it's confusing):

    $ ./BUILD/util/hb-shape --font-file NotoColorEmoji.ttf --unicodes 1f1f9,1f1e6 --direction rtl
    [gid1228=0+2550]

When Blink processes a mixture of Hebrew and regional-indicator codes, does it in fact apply RTL direction to the regional-indicators? AFAICS that's what would cause harfbuzz to return a different flag ligature, not the Hebrew script tag.

**Comment by arenevier** [comment]
Yes, the problem happens when script is hebrew *and* direction is ltr.

(./BUILD/util/hb-shape --font-file NotoColorEmoji.ttf --unicodes 1f1f9,1f1e6 --direction rtl will set the script to hebrew because `hb_buffer_guess_segment_properties()` is called)

**Comment by dscorbett** [comment]
See #1006.

**Comment by jfkthame** [comment]
Ah, right -- sorry, I was forgetting that `--script hebr` would imply `--direction rtl` by default. So yes, this does make sense.

I'm still unclear whether we ought to change harfbuzz behavior here or not, as discussed above; there seem to be arguments that could be made either way.

**Comment by khaledhosny** [comment]
> The discussion at [crbug.com/1054512#c18](https://crbug.com/1054512#c18) discovered that Gecko runs script segmenter for each bidi runs, while Blink runs for the entire paragraph. For this test case, this is reason why the regional indicators are tagged as Hebrew only in Blink.
> 
> I'm fine to match this behavior to Gecko, if people here can reach consensus that it is the best way to go. It solves the reported problem, though doesn't really answer the directional override issue Jonathan mentioned.

Doing script itemization on individual script runs is suboptimal, as digits in RTL text will result in multiple bidi runs and the digits will not get Arabic script as they should, also they might cause punctuation only runs which also will not get Arabic script. See https://bugzilla.mozilla.org/show_bug.cgi?id=759284.

**Comment by behdad** [comment]
There's run direction, and then there's overrides.  Overrides are a weird beast.  I'm happy with a solution that breaks with overrides if have to.

I'm of the opinion that Unicode wants these pairs to be considered atomic, and obviously encoded in logical order.  For that reason, I think even an RTL override should NOT change the meaning of the flag.  Ie. form the same flag.  For these reasons, I believe we *should* do something like what Khaled prototyped.

When a flag is NOT formed by the font, that is considered incomplete font.  So I wouldn't particularly worry if the two individual characters render next to each other in the wrong direction in RTL.  However, if we do insist that they render as individual letters correctly as well, we can do another change such that during reversing of the run, we reverse such sequences back.  We don't do that for mark attachments, etc, so that would be additional code / change.  Or as a very ugly hack we can just change the positioning of those two.  I'm not very comfortable with this.

I haven't researched what Unicode categorizes a pair of these that doesn't define a flag.  Is that invalid text?  Does Unicode specify how it should render?  I know in Noto Color Emoji, @dougfelt was writing GSUB rules to position the two letters in a "missing flag" glyph.  To do that correctly font developer needs to use `rlta` feature to differentiate RTL direction and position them properly.

In conclusion, I think something simple like Khaled's patch is appropriate here.  I'm surprised we didn't do this before.

**Comment by dougfelt** [comment]
It's been awhile since I worked on Google's emoji font.

I don't know that Unicode says one way or the other whether these pairs should be atomic in the face of directional overrides. Regional indicators characters are bidi L so in embeddings they would retain their LTR ordering (and break directional runs).  Since there are inverse pairs that represent different flags it's probably important to keep them LTR in an embedding but I think all bets are off in an override.

I'm not quite sure what Google's emoji font does these days. The preferred approach is to show a missing flag if the font does not support a ligature for the chars.  I did this for subregion flags (which use tag characters, not RI characters, and so are BN not L, so the situation is not quite the same).  I don't remember if I just let it fallback to chars (no missing flag glyph) if the RI sequence wasn't supported, I might have had a default gsub rule that replaced all unsupported RI sequences with the missing flag glyph.  The approach I took for emoji sequences containing zwj was to form ligatures given either ordering (forward or reverse), but I only encoded the logical order for RI pairs and subregions.  If this is still the case then subregion flags would get mangled in an embedding, I can't remember.

**Comment by behdad** [comment]
Humm.  That does match my memory. If they are bidi=L, why do multiple flags in a sequence not always line up left to right even in the middle of Arabic text?  Puzzled.

**Comment by jfkthame** [comment]
They do, surely? The flags in https://codepen.io/jfkthame/pen/gOpKVpr appear in LTR order (AU, AD, DE) in the midst of Arabic text.

**Comment by jfkthame** [comment]
> I'm of the opinion that Unicode wants these pairs to be considered atomic, and obviously encoded in logical order. For that reason, I think even an RTL override should NOT change the meaning of the flag. Ie. form the same flag. For these reasons, I believe we _should_ do something like what Khaled prototyped.

My concern is that I can't find any support for that in UAX #9, and without explicit support from the Unicode standard, I think conformity to the normative algorithm (and hence interoperability with other systems that strictly follow the same standard) trumps other considerations for an edge-case like this.

**Comment by behdad** [comment]
> They do, surely? The flags in https://codepen.io/jfkthame/pen/gOpKVpr appear in LTR order (AU, AD, DE) in the midst of Arabic text.

Donno why Android is showing otherwise.  Weird.  cc @nona-google

**Comment by jfkthame** [comment]
That seems non-Unicode-conformant AFAICS. It'd be nicer if they were neutral, perhaps, but that's not what the standard says, and reliable interoperability requires that we all follow the same standard.

(A user wanting a sequence of flags in an RTL paragraph to appear in RTL order could do something like inserting RLM in between them to ensure this. A bit of a pain, but it should be reliable. But if systems disagree about the bidi behavior, chaos will ensue...)

**Comment by nona-google** [comment]
(I don't follow the all conversation here)
The Android text stack does Bidi run segmentation first, then font itemization, then finally script segmentation. I think that's the reason why Android text shows flags in LTR order even in the middle of the RTL text.

**Comment by arenevier** [comment]
Is there a consensus on whether it would be better be fixed on harfbuzz side, or on or chrome side?

**Comment by behdad** [comment]
I think the consensus is that  since RI characters have bidi=L, no action is needed in HarfBuzz.

**Comment by behdad** [comment]
> I think the consensus is that since RI characters have bidi=L, no action is needed in HarfBuzz.

Humm.  But if client marks them as script=Hebrew or other RTL script, we're going to flip them to apply ligatures and that fails unless font has been built with that in mind.  So it's not a completely non-issue.  Let's think more.

**Comment by khaledhosny** [comment]
I see that Chrome opted to replicate the Firefox limitation of doing script segmentation on individual bidi runs 😥

**Comment by behdad** [comment]
Any volunteers to write to Unicode Consortium and propose updating the Bidi algorithm to treat RI pairs as atomic, and also make the atomic results bidi-neutral so they accumulate in the run direction instead of always running LTR next to each other?

**Comment by kojiishi** [comment]
I agree that there may be something Unicode can help, but let me try to understand the problem a little more.

If I understand the problem correctly, this issue happens to all ligatures of script=COMMON characters (and if the renderer segments script across bidi-runs,) correct? I quickly tested Emoji ligatures but they seem to work fine with `bidi-override`:
https://jsbin.com/kifaqit/edit?html,output
and the behavior looks consistent between Gecko, Blink, and WebKit.

I don't know from top of my head what we do for Emoji ligatures, but does applying the same logic makes sense?

**Comment by behdad** [comment]
I asked Mark Davis for insight. This is his quick response (after acknowledging the pain and mistake that is RI):

> The semantics of the RI pairs are that each pair, starting in memory from first found to last, are treated as atomic units. This is independent of what the layout direction is. So if X, Y, and Z are RI, and we have a sequence of code points in memory:
> 
> abcXYZdef
> 
> Then the grouping is  
> 
> a, b, c, XY, Z, d, e, f,
> 
> And XY refers to the country code  XY, never YX— even under RLO (or equivalent higher-level protocols).
> 
> I know this is an awful pain in the ass to work with, even without the BIDI component.
> 
> Mark

If we accept that RI pairs *must* ligate in correct rendering, then there's no apparent violation if we cluster them together. Is only when they don't ligate that we will be violating RLO-enforced RI pairs...

For the sake of doing the right thing in most common cases, I suggest we implement the pairing in our clustering algorithm.

**Comment by drott** [comment]
>  I suggest we implement the pairing in our clustering algorithm.

+1, I am happy with that, but curious to hear @jfkthame's opinion, too.

**Comment by khaledhosny** [comment]
> The discussion at [crbug.com/1054512#c18](https://crbug.com/1054512#c18) discovered that Gecko runs script segmenter for each bidi runs, while Blink runs for the entire paragraph. For this test case, this is reason why the regional indicators are tagged as Hebrew only in Blink.
> 
> I'm fine to match this behavior to Gecko, if people here can reach consensus that it is the best way to go. It solves the reported problem, though doesn't really answer the directional override issue Jonathan mentioned.

@kojiishi With this fixed in HarfBuzz, I think Blink should revert back to run the script segmener on the entire paragraph.
